tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,357 @@
+  }
+
+  std::vector<std::string> tokens = tokenize(path);
+  FSNode* cur = root_;
+  for (unsigned long i = 0; i < tokens.size() - 1; i++) {
+    auto token = tokens[i];
+    if (!cur->has_child(token)) {
+      cur->children_[token] = new Directory(token);
+    }
+    cur = cur->children_[token];
+  }
+
+  std::string filename = tokens[tokens.size() - 1];
+  cur->children_[filename] = new File(filename);
+  *new_node = cur->children_[filename];
+
+  return Status::Ok();
+}
+
+Status MemFilesystem::write(
+    const std::string& path, const void* data, const uint64_t nbytes) {
+  FSNode* node;
+  RETURN_NOT_OK(get_node(path, &node));
+  /* If the file doesn't exist, create it */
+  if (node == nullptr) {
+    touch(path, &node);
+    /* If it's a directory, ignore */
+  } else if (node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(std::string(
+        "Failed to write on directory, write is only possible on files.")));
+  }
+
+  return ((File*)node)->append(data, nbytes);
+}
+
+Status MemFilesystem::read(
+    const std::string& path,
+    const uint64_t offset,
+    void* buffer,
+    const uint64_t nbytes) const {
+  FSNode* node;
+  RETURN_NOT_OK(get_node(path, &node));
+  if ((node == nullptr) || node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(
+        std::string("File not found, read failed for : " + path)));
+  }
+
+  return ((File*)node)->read(offset, buffer, nbytes);
+}
+
+std::vector<std::string> MemFilesystem::tokenize(
+    const std::string& path, const char delim) {
+  std::vector<std::string> tokens;
+  std::stringstream ss(path);
+  std::string token;
+  while (std::getline(ss, token, delim)) {
+    if (!token.empty()) {
+      tokens.emplace_back(token);
+    }
+  }
+
+  return tokens;
+}
+
+Status MemFilesystem::get_node(const std::string& path, FSNode** node) const {
+  assert(node);
+  if (root_ == nullptr) {
+    return LOG_STATUS(Status::MemFSError(
+        std::string("Directory creation not possible, in-memory filesystem not "
+                    "initialized")));
+  }
+
+  std::vector<std::string> tokens = tokenize(path);
+  auto cur = root_;
+  for (const auto& token : tokens) {
+    if (cur->has_child(token)) {
+      cur = cur->children_[token];
+    } else {
+      *node = nullptr;
+      return Status::Ok();
+    }
+  }
+
+  *node = cur;
+  return Status::Ok();
+}
+
+void MemFilesystem::delete_rec(FSNode* node) {
+  for (auto child : node->children_) {
+    delete_rec(child.second);
+  }
+  delete node;
+}
+
+Status MemFilesystem::remove(const std::string& path) {
+  FSNode* node;
+  RETURN_NOT_OK(get_node(path, &node));
+  if (node == nullptr) {
+    return LOG_STATUS(Status::MemFSError(
+        std::string("File not found, remove failed for : " + path)));
+  }
+
+  std::vector<std::string> tokens = tokenize(path);
+  auto cur = root_;
+  FSNode* parent = nullptr;
+  for (const auto& token : tokens) {
+    parent = cur;
+    cur = cur->children_[token];
+  }
+
+  delete_rec(cur);
```

Comment:
As I mentioned earlier, we could eliminate this logic by stack-allocating nodes. If we think we will have enough nodes to overflow the stack memory, we should go ahead and use unique_ptr or shared_ptrs to manage the FSNode objects.

Reply:
I was not sure about the use case of the memory filesystem, will the stack be enough? Let me know and I will be happy to go with whatever you choose.
That said, you are right about unique_ptrs, old school C programmer here went the unsafe way once again :)))

Reply:
Let's go with storing `FSNode*` objects in the `children_` maps (and the `root_`) as `unique_ptr<FSNode>`. This requires a little bit of context:

1. I strongly prefer `unique_ptr` to `shared_ptr` because it makes the ownership/lifetime rules more obvious. In my experience, it is difficult to order destructors (and their cleanup of unmanaged resources) when a bunch of `shared_ptrs` are moving around.
2. By using a `unique_ptr` we will have to pass a raw c-pointer to other objects that do not manage the FSNode lifetimes. This is fine, because the syntax makes it clear that a raw pointer is not managed by whoever is referencing it.

So for example, in my other code snippet where I invoked `lookup_node(FSNode* current_node, ..., ...)`, I would call it from the root as:
```
// `root_` is of type `unique_ptr<FSNode>`
FSNode* child = lookup_node(root_.get(), ..., ...);
// we do not own `child`, it is a `unique_ptr` within an `unordered_map<string, unique_ptr<FSNode>> FSNode::children_`
```

Reply:
Cool thanks for the detailed comment. I tried to use unique_ptrs instead, but please double check if I did it right 😊